### PR TITLE
Topic recognition: (46445010) Task 7: VQVAE for HIPMRI prostate cancer image reconstruction.

### PR DESCRIPTION
**Pull request by esdafinesseda**
Author: George Reid-Smith

Project: Implements, trains, and runs inference on a VQVAE model that reconstructs HIPMRI prostate cancer images through encoding, quantization, and decoding.

The model training is designed to encode input images into a high and low layer, then match encoded variables at these layers  to vectors in a learned discrete codebook, before reconstructing the image with using these vectors with decoders. This project provides the details on how to train and run inference on the model and provides a model training example that reconstructs images with structured similarity of ~75%. 

Several suggested improvements have been included to further the performance of the model.

Structure:
`README.md` - project outline, model architecture, execution instructions, example inputs and outputs, further work recommendations.

`config/` - configuration files for model parameters and directories.
`logging/` - outputs of training and inference.
`models/` - best model (up to SSIM) after a training execution.
`samples/` - generated images from training and inference.
`visuals/` - model architecture figures and training plots.

`dataset.py` - Custom Dataset and Dataloader classes for the HIPMRI dataset.
`metrics.py` - Methods for loss and ssim calculations.
`modules.py` - All torch modules for the VQVAE.
`predict.py` - Script to run inference on the model.
`train.py` - Script to run model training.
`utils.py` - General utilities for setup, configuration, and saving outputs.

`requirements.txt` - dependency versions for the project.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear and resembling reconstructed images, with an SSIM over 0.7 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

3. Design:
Highly modularized design and good programming practice using utils script (1 Mark)

4. Commenting:
Sufficient docstrings and detailed comments (1 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- You could be a bit more specific in describing the training process by including some key hyperparameters and tuning ---- similar to how you justified the batch size (e.g. why you ended up choosing LR of 3e-5, what else you had attempted to improve the performance and what were their implications);
- Images not shown properly on README.md due to incorrect path configuration;
- Very good project!

**Comment by YashMitttal** [comment]
Your report covers all necessary aspects of the VQVAE implementation, training process, and evaluation. You've made good decisions on model architecture, pre-processing, and reproducibility.

A small area of improvement could be around your pull request. It can be more descriptive about the project's specific details. But good job overall.

**Comment by esdafinesseda** [comment]
Provided greater detail on the "details of the project" as recommended.

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by shakes76** [comment]
Good work!
